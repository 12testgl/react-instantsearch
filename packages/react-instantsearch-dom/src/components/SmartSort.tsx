import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { createClassNames } from '../core/utils';

const cx = createClassNames('SmartSort');

type ComponentProps = {
  isSmartSorted: boolean;
};

export type SmartSortComponent = {
  buttonTextComponent: React.FunctionComponent<ComponentProps>;
  className?: string;
  isVirtualReplica: boolean;
  isSmartSorted: boolean;
  textComponent: React.FunctionComponent<ComponentProps>;
  refine(relevancyStrictness: number | undefined): void;
};

const SmartSort: React.FC<SmartSortComponent> = ({
  buttonTextComponent: ButtonTextComponent,
  className = '',
  isVirtualReplica,
  isSmartSorted,
  textComponent: TextComponent,
  refine,
}) => {
  if (!isVirtualReplica) {
    return null;
  }

  return (
    <div className={classNames(cx(''), className)}>
      <div className={cx('text')}>
        <TextComponent isSmartSorted={isSmartSorted} />
      </div>
      <button
        className={cx('button')}
        onClick={() => refine(isSmartSorted ? 0 : undefined)}
      >
        <ButtonTextComponent isSmartSorted={isSmartSorted} />
      </button>
    </div>
  );
};

SmartSort.propTypes = {
  buttonTextComponent: PropTypes.func.isRequired,
  className: PropTypes.string,
  isVirtualReplica: PropTypes.bool.isRequired,
  isSmartSorted: PropTypes.bool.isRequired,
  refine: PropTypes.func.isRequired,
  textComponent: PropTypes.func.isRequired,
};

export default SmartSort;
