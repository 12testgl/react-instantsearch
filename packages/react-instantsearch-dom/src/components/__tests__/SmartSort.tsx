import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import SmartSort from '../SmartSort';

const textComponent = ({ isSmartSorted }) => (
  <div>
    {isSmartSorted
      ? 'We are filtering unrelevant results'
      : 'We are not filtering unrelevant results'}
  </div>
);

const buttonTextComponent = ({ isSmartSorted }) => (
  <div>{isSmartSorted ? 'See all results' : 'See relevant results'}</div>
);

textComponent.propTypes = {
  isSmartSorted: PropTypes.bool.isRequired,
};

buttonTextComponent.propTypes = {
  isSmartSorted: PropTypes.bool.isRequired,
};

describe('SmartSort', () => {
  it("returns null if it's not a virtual replica", () => {
    const tree = renderer.create(
      <SmartSort
        buttonTextComponent={buttonTextComponent}
        isVirtualReplica={false}
        isSmartSorted={false}
        refine={() => {}}
        textComponent={textComponent}
      />
    );

    expect(tree.toJSON()).toBeNull();
  });

  it('accepts a custom className', () => {
    const tree = renderer.create(
      <SmartSort
        buttonTextComponent={buttonTextComponent}
        className="MyCustomSmartSort"
        isVirtualReplica={true}
        isSmartSorted={false}
        refine={() => {}}
        textComponent={textComponent}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('forward isSmartSorted to props components', () => {
    const tree = renderer.create(
      <SmartSort
        buttonTextComponent={buttonTextComponent}
        isVirtualReplica={true}
        isSmartSorted={true}
        refine={() => {}}
        textComponent={textComponent}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
