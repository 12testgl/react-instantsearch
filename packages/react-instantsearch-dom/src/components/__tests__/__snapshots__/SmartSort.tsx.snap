// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartSort accepts a custom className 1`] = `
<div
  className="ais-SmartSort MyCustomSmartSort"
>
  <div
    className="ais-SmartSort-text"
  >
    <div>
      We are not filtering unrelevant results
    </div>
  </div>
  <button
    className="ais-SmartSort-button"
    onClick={[Function]}
  >
    <div>
      See relevant results
    </div>
  </button>
</div>
`;

exports[`SmartSort forward isSmartSorted to components 1`] = `
<div
  className="ais-SmartSort"
>
  <div
    className="ais-SmartSort-text"
  >
    <div>
      We are filtering unrelevant results
    </div>
  </div>
  <button
    className="ais-SmartSort-button"
    onClick={[Function]}
  >
    <div>
      See all results
    </div>
  </button>
</div>
`;
